### PR TITLE
add dmz url

### PR DESCRIPTION
**Pull request by witmicko**
## JIRA 
https://issues.jboss.org/browse/AEROGEAR-2570
## Changes
- [x] added `mobile set --client=<clientid> --name=<valueName> --value=<value>` command
- [x] added `mobile set client <clientName> --patch=<JSON Patch>`
- [x] unit tests 

Notes:
- `oc patch` https://docs.openshift.com/container-platform/3.3/cli_reference/basic_cli_operations.html#patch
 
## Verification notes
Note: Its best to verify this with https://github.com/aerogear/mobile-core/pull/73#pullrequestreview-116551997
- checkout this branch locally
- build
- add mobile crd as described on pinned PR
- `./mobile set value -c=myapp-android -n=dmzUrl -v=www.ddd.ssd --namespace=myproject`
- `./mobile set client myapp-android --patch='{"spec": {"dmzUrl": "www.dfgdfgrldorder.com"}}' --namespace=myproject -o=json`
- `./mobile set client myapp-android --patch='{"spec": {"dmzUrl": null}}' -namespace=myproject -o=json`
- `./mobile set client myapp-android --patch='{"spec": {"dmzUrl": "www.anotherexample.com"}}' --namespace=myproject -o=json`
- `./mobile set value -c=myapp-android -n=dmzUrl  --namespace=myproject -o=json` should set it to ""
- `./mobile set value -c=myapp-android -n=dmzUrl -v=empty.com --namespace=myproject -o=json`
- `./mobile set value -c=myapp-android -n=dmzUrl -v=null --namespace=myproject -o=json`

### CODE REVIEW

File: pkg/cmd/clients.go [hunk 1]
```diff
@@ -226,7 +227,7 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
 				return cmd.Usage()
 			}
 			clientID := args[0]
-
+			fmt.Println(args)
```
Comment (maleck13) [inline pkg/cmd/clients.go]: can we remove this

File: pkg/cmd/clients.go [hunk 2]
```diff
@@ -241,3 +242,53 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+	command := &cobra.Command{
+		Use:   "client <clientID>",
+		Short: "Patches mobileclient with provided json",
+		Long: `set client allows you to patch a mobile client in your namespace.
+			   Run the "mobile get clients" command from this tool to get the client ID.`,
+		Example: `mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}' --namespace=myproject 
+  			      kubectl plugin mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}'
+				  oc plugin mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}'`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var err error
```
Comment (maleck13) [inline pkg/cmd/clients.go]: can declare these as 
```
var(
 err error
  res *v1alpha1.MobileClient
   ns string
)
```

File: artifacts/mobileclient_crd.yaml [hunk 3]
```diff
@@ -29,4 +29,7 @@ spec:
               pattern: '([\w-])'
             appIdentifier:
               type: string
-              pattern: '([\w-])'
\ No newline at end of file
+              pattern: '([\w-])'
+            dmzUrl:
+              type: string
+              pattern: '(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)'
```
Comment (philbrookes) [inline artifacts/mobileclient_crd.yaml]: This update needs to be reflected here too:
https://github.com/aerogear/mobile-core/blob/master/installer/roles/create-mobile-client-crd/files/mobile-client-crd.yaml

File: pkg/cmd/clients.go [hunk 4]
```diff
@@ -241,3 +242,53 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+	var patch string
+	command := &cobra.Command{
+		Use:   "client <clientID>",
+		Short: "Patches mobileclient with provided json",
+		Long: `set client allows you to patch a mobile client in your namespace.
+			   Run the "mobile get clients" command from this tool to get the client ID.`,
+		Example: `mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}' --namespace=myproject 
+  			      kubectl plugin mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}'
+				  oc plugin mobile set client <clientID> --patch='{"spec": {"dmzUrl": "www.dmz.com"}}'`,
+		RunE: func(cmd *cobra.Command, args []string) error {
```
Comment (philbrookes) [inline pkg/cmd/clients.go]: Let's add some unit tests for this new command as well.

File: artifacts/mobileclient_crd.yaml (hunk 3 above)
Comment (witmicko) [reply artifacts/mobileclient_crd.yaml]: @philbrookes https://github.com/aerogear/mobile-core/pull/73

File: pkg/cmd/clients_test.go [hunk 5]
```diff
@@ -544,3 +544,244 @@ func TestMobileClientsCmd_TestCreateClient(t *testing.T) {
...
+				if nil == client {
+					t.Fatal("expected a mobile client but got nil")
+				}
+				if client.Spec.Name != "my-new-app" {
+					t.Fatalf("expected an app with name %s but got %s ", "myapp", client.Spec.Name)
+				}
+			},
+		},
+		{
+			Name: "Test set with name data",
```
Comment (philbrookes) [inline pkg/cmd/clients_test.go]: this test has the same name as the previous test and appears to be doing the same thing?

File: pkg/cmd/clients_test.go [hunk 6]
```diff
@@ -544,3 +544,244 @@ func TestMobileClientsCmd_TestCreateClient(t *testing.T) {
...
+		ErrorPattern string
+		ExpectUsage  bool
+		ClientId     string
+		ValueName    string
+		Value        string
+		Flags        []string
+		Validate     func(t *testing.T, c *v1alpha1.MobileClient)
+	}{
+		{
+			Name: "Test set dmz url",
```
Comment (philbrookes) [inline pkg/cmd/clients_test.go]: Can we add a test with a URL that fails that regex too?

File: pkg/cmd/clients.go [hunk 7]
```diff
@@ -241,3 +241,121 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+				patch = fmt.Sprintf("{\"spec\": {\"%s\": null}}", name)
+			} else {
+				patch = fmt.Sprintf("{\"spec\": {\"%s\": \"%v\"}}", name, value)
+			}
+
+			res, err = cc.mobileClient.MobileV1alpha1().MobileClients(ns).Patch(clientId, types.MergePatchType, []byte(patch))
+			if err != nil {
+				return errors.Wrap(err, "failed to set value in mobile client with clientID "+clientId)
+			}
+			//outType := outputType(cmd.Flags())
```
Comment (vchepeli) [inline pkg/cmd/clients.go]: remove commented like

File: pkg/cmd/clients.go [hunk 8]
```diff
@@ -241,3 +241,121 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+			if err := cc.Out.Render(cmd.Name(), "json", res); err != nil {
+				return errors.Wrap(err, fmt.Sprintf(output.FailedToOutPutInFormat, "mobile client", "json"))
+			}
+
+			return nil
+		},
+	}
+	command.PersistentFlags().StringVarP(&clientId, "client", "c", "", "value")
+	command.PersistentFlags().StringVarP(&valueName, "name", "n", "", "value")
```
Comment (vchepeli) [inline pkg/cmd/clients.go]: `command.PersistentFlags().StringVarP(&valueName, "name", "n", "", "name")`

File: pkg/cmd/clients.go [hunk 9]
```diff
@@ -241,3 +241,121 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+			var patch string
+			if value == "null" || value == "" {
+				patch = fmt.Sprintf("{\"spec\": {\"%s\": null}}", name)
+			} else {
+				patch = fmt.Sprintf("{\"spec\": {\"%s\": \"%v\"}}", name, value)
+			}
+
+			res, err = cc.mobileClient.MobileV1alpha1().MobileClients(ns).Patch(clientId, types.MergePatchType, []byte(patch))
+			if err != nil {
+				return errors.Wrap(err, "failed to set value in mobile client with clientID "+clientId)
```
Comment (vchepeli) [inline pkg/cmd/clients.go]: You can use `fmt.Sprintf("failed to set value in mobile client with clientID %s", clientID)`

File: pkg/cmd/clients.go [hunk 10]
```diff
@@ -241,3 +241,121 @@ Run the "mobile get clients" command from this tool to get the client ID.`,
...
+			}
+
+			ns, err = currentNamespace(cmd.Flags())
+			if err != nil {
+				return errors.Wrap(err, "failed to get namespace")
+			}
+
+			res, err = cc.mobileClient.MobileV1alpha1().MobileClients(ns).Patch(clientID, types.MergePatchType, []byte(patch))
+			if err != nil {
+				return errors.Wrap(err, "failed to patch mobile client with clientID "+clientID)
```
Comment (vchepeli) [inline pkg/cmd/clients.go]: `fmt.Sprintf("failed to set value in mobile client with clientID %s", clientID)`

**Pull request merged**